Add Stretch configuration to the Storage Cluster

raghavendra-talur opened the pull request:
This PR adds the arbiter mode of deployment to the Storage Cluster.

Two new fields are added to the Storage Cluster

* ArbiterSpec
* NodeTopologies

If ArbiterSpec.Enable is set to true *and* if
NodeTopologies.ArbiterLocation is set then the Ceph Cluster is
configured with arbiter enabled.

If ArbiterSpec.Enable is true and no NodeTopologies.ArbiterLocation is
provided then the Storage Cluster fails to reconcile.

If ArbiterSpec.Enable is false then NodeTopologies.ArbiterLocation is
ignored even if set.

The arbiter mode differs from the non-arbiter mode in the following
ways:

* The minimum and the maximum number of MONs are 5 instead of 3.
* The OSDs run only in 2 of the 3 failure domains. The third failure
  domain is only used to host a MON pod.
* Mininum number of OSDs in each failure domain is 2 instead of 1.
* The CRUSH Map is configured in replica 4 configuration instead of 3.
* The OSDs can be scaled in multiples of 2 and not 3.
* The placement algorithm doesn't restrict the arbiter MON pod to nodes
  which have the default/user-provided OCS node label. Rest of the ceph
  pods have the nodeAffinity same as in the non-arbiter mode.
* The arbiter MON pod is scheduled to run on any of the nodes in the
  ArbiterLocation in the failure domain.
* The arbiter MON pod has toleration for the master node taint by default. It
  can be disabled by setting ArbiterSpec.DisableMasterNodeToleration to true.

ocs-operator's placement strategy for non-arbiter mode is to select a failure
domain and then spread the ceph resources across the failure domain.

Keeping with the same principle, ocs-operator determines the failure domain
automatically in the the arbiter mode too and tries to match one of the failure
domain locations to the NodeTopologies.ArbiterLocation.

In both the modes, it might sometimes be desirable that the user gets to
specify the failure domain through the Storage Cluster and not rely on the
automatic detection. They may also wish to specify the node topology
themselves.  For this reason, we did not add the ArbiterLocation field to the
ArbiterSpec and added it to the TopologyMap struct instead.

Work needs to be done to reconcile StorageClusterSpec.FlexibleScaling,
StorageClusterSpec.NodeTopologies.FailureDomain, and
StorageClusterStatus.FailureDomain to let users choose the node topology and
the failure domain. This can be done in the future, we have mad

raghavendra-talur commented:
/hold

raghavendra-talur commented on pkg/apis/ocs/v1/storagecluster_types.go:
```diff
@@ -155,6 +156,16 @@ type EncryptionSpec struct {
+// Arbiter mon pods don't require the nodes to have the ocs node label.
+type ArbiterSpec struct {
+	// +optional
+	FailureDomainValue string `json:"failureDomainValue,omitempty"`
```
@jarrpa @umangachapagain @obnoxxx @iamniting Although I still need to work on the placement code for arbiter, I would like to get your opinion on this change in the API.

Rook needs both the failure domain name and the 3 values for that failure domain to configure the stretch cluster. ocs-operator determines the failure domain name internally and therefore requires only the value of the failure domain that will be arbiter.

Is `FailureDomainValue` an acceptable name ? If not, any suggestions?

raghavendra-talur replied:
Here is the change in Rook https://github.com/rook/rook/pull/6489/commits/b91f4211c9833f89b107599cd6fd091f4c7f06f2#diff-23d4ca6ed288cf0c307912d860846319f62eb5f60846961916c0395a82d76cfaR123

raghavendra-talur replied:
If I am not wrong, ocs-operator has been determining the failure domain itself. The UI too, for the arbiter config is going to pass only the value of the failure domain(for example us-east-2b).

IIUC,  in the UI, the failure domain is always shown as a zone.

raghavendra-talur replied:
Based on @jarrpa's and @travisn's comments, the name of the field is now ArbiterLocation, this is the input given to the ocs-operator in the StorageCluster CR and is just the value of the failure domain that is for arbiter. The failure domain label that is passed to rook is a complete label.

umangachapagain replied:
> Based on @jarrpa's and @travisn's comments, the name of the field is now ArbiterLocation
>
Sounds good, but where is this comment?

raghavendra-talur commented:
/unhold

raghavendra-talur commented on pkg/deploy-manager/arbiter.go:
```diff
@@ -0,0 +1,67 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// electArbiterZone is a helper function just for the tests. In real deployments, the arbiter zone will be picked by the user.
```
This code belongs to a different package in the ocs-operator, deploy-manager which is sort of a replacement/alternative to the openshift console. Therefore is it not prefixed with `test`. It is not compiled into the ocs-operator binary though.

raghavendra-talur commented on pkg/deploy-manager/arbiter.go:
```diff
@@ -0,0 +1,67 @@
+// electArbiterZone is a helper function just for the tests. In real deployments, the arbiter zone will be picked by the user.
+// In our tests, we will just pick the zone with the least number of nodes as the arbiter zone.
+func (t *DeployManager) electArbiterZone() (arbiterZoneElect string, err error) {
+	nodes, err := t.k8sClient.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: "node-role.kubernetes.io/worker"})
```
Yes, I am making an assumption for the deploy-manager that the arbiter pods will be deployed on a worker node. This is to keep the deploy manager simple. The placement algorithm in the ocs-operator now has the master node toleration set for the arbiter pods.

raghavendra-talur commented on pkg/deploy-manager/config.go:
```diff
@@ -27,8 +27,15 @@ const DefaultStorageClusterName = "test-storagecluster"
+
+func (t *DeployManager) getMinOSDsCount() int {
+	if t.IsArbiterEnabled() {
+		return minOSDsCountArbiter
```
Yes. The above scenario is covered; added in the `placement` related commit.

raghavendra-talur commented:
@jarrpa @travisn @obnoxxx @umangachapagain @iamniting I believe this is PR is now ready for review.

I have provided design details in the description of the PR.

obnoxxx commented:
fyi, functest was failing like this before the update:
```
Failure [21.196 seconds]
[BeforeSuite] BeforeSuite 
/go/src/github.com/openshift/ocs-operator/functests/ocs/ocs_test.go:17
  Expected
      <*errors.errorString | 0xc00032a230>: {
          s: "No arbiter nodes found, we need atleast 1",
      }
  to be nil
  /go/src/github.com/openshift/ocs-operator/functests/setup_teardown.go:17
------------------------------ 
```

@raghavendra-talur - what did you change regarding this in the last update?

raghavendra-talur commented:
> fyi, functest was failing like this before the update:
> 
> ```
> Failure [21.196 seconds]
> [BeforeSuite] BeforeSuite 
> /go/src/github.com/openshift/ocs-operator/functests/ocs/ocs_test.go:17
>   Expected
>       <*errors.errorString | 0xc00032a230>: {
>           s: "No arbiter nodes found, we need atleast 1",
>       }
>   to be nil
>   /go/src/github.com/openshift/ocs-operator/functests/setup_teardown.go:17
> ------------------------------ 
> ```
> 
> @raghavendra-talur - what did you change regarding this in the last update?

The last update only added csv files that are generated when `make gen-latest-csv` is run.

Regarding the functest failure, the arbiter mode of functest expects 5 worker nodes. I will edit it such that it can run with the existing infra and/or I will update here after editing the functest ci job to provide more worker nodes.

umangachapagain reviewed (changes requested):
Commit - `deploy: add files updates by 'make gen-latest-csv'` should have changes related to Rook CRDs only. It seems there are some unwanted changes from a later commit.

umangachapagain commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -343,8 +345,17 @@ func newExternalCephCluster(sc *ocsv1.StorageCluster, cephImage string, monitori
+func getMinNodes(sc *ocsv1.StorageCluster) int {
+
+	if arbiterEnabled(sc) {
+		// We need a minimum of 2 OSDs in each of the two zones when arbiter is enabled
+		return defaults.ArbiterModeDeviceSetReplica * 2
```
`defaults.ArbiterModeDeviceSetReplica` is already 2. Why do we need to multiply by 2?

obnoxxx replied:
quoting from the definition:
```
ArbiterModeDeviceSetReplica is the default number of Rook-Ceph StorageClassDeviceSets per StorageCluster StorageDeviceSet when arbiter is enabled
```

In normal mode we have `3`here.
Now, we want 2 replicas per zone. (So that we have two copies in each zone.) 
But I think we want to make the constant factor of `2`here configurable too.
I think for this release it might be ok, but later, we want to be able to do one copy per zone...

raghavendra-talur replied:
Good question:
There are two equations that we need
`minDevices  == minNodes == replicaPerFailureDomain * numFailureDomains`
`minAdditionalDevicesForExpansion == minAdditionalNodesPerExpansion == numFailureDomains`

1. In the non-arbiter cases, the minimum requirement is replica 1 within the failure domain and 3 failure domains in the cluster. To expand the cluster, you can start with the addition of 1 OSD per failure domain. Therefore,
`minDevices  == minNodes == 1 * 3`
`minAdditionalDevicesForExpansion == minAdditionalNodesPerExpansion == 3`

2. In the arbiter case, however, the minimum requirement is replica 2 within the failure domain and 2 failure domains in the cluster. To expand the cluster, you can start with the addition of 1 OSD per failure domain. Therefore,
`minDevices  == minNodes == 2 * 2`
`minAdditionalDevicesForExpansion == minAdditionalNodesPerExpansion == 2`

With that explanation, and to incorporate what @obnoxxx suggested, I will update the code to do the following:
1. add a default for ReplicaPerFailureDomain for arbiter and non arbiter configurations
2. add functions that capture the above equations

raghavendra-talur replied:
Updated

umangachapagain commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -564,3 +575,43 @@ func (r *ReconcileStorageCluster) checkTuneStorageDevices(ds ocsv1.StorageDevice
 	return diskSpeedUnknown, nil
 }
+
+func generateStretchClusterSpec(sc *ocsv1.StorageCluster) *cephv1.StretchClusterSpec {
```
```suggestion
func generateStretchClusterSpec(sc *ocsv1.StorageCluster) cephv1.StretchClusterSpec {
```

raghavendra-talur replied:
cephv1.MonSpec requires *cephv1.StretchClusterSpec. It just makes it easy to match with what Cephv1 defines it as.

jarrpa replied:
I think this is fine in this case.

umangachapagain commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -564,3 +575,43 @@ func (r *ReconcileStorageCluster) checkTuneStorageDevices(ds ocsv1.StorageDevice
+		Arbiter: true,
+	})
+
+	return &stretchClusterSpec
```
```suggestion
	return stretchClusterSpec
```

raghavendra-talur replied:
cephv1.MonSpec requires *cephv1.StretchClusterSpec. It just makes it easy to match with what Cephv1 defines it as.

jarrpa replied:
I think this is fine in this case.

umangachapagain commented on pkg/controller/storagecluster/reconcile.go:
```diff
@@ -190,6 +190,13 @@ func (r *ReconcileStorageCluster) reconcilePhases(
 
+	if err := validateArbiterSpec(instance, reqLogger); err != nil {
+		instance.Status.Phase = statusutil.PhaseError
+		r.client.Update(context.TODO(), instance)
```
Error needs to be handled.

iamniting replied:
I am doing this intentionally, I just want to update it, if it does not update it I don't care much about it.
Moreover, If we are going to handle update err then we will miss arbiter err and I do not want to club them together.

umangachapagain replied:
Well this approach isn't going to make the linter happy.

I'd suggest this if your intent is to ignore the error
```suggestion
		_ = r.client.Update(context.TODO(), instance)
```

obnoxxx replied:
Agreed. Be explicit (by using `_ :=`when ignoring return values, in particular errors.

iamniting replied:
Done

obnoxxx commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -422,7 +444,7 @@ func newStorageClassDeviceSets(sc *ocsv1.StorageCluster, serverVersion *version.
 			// to be updated.
 			// TODO: Remove this behavior when OCP console is updated
-			count = count / 3
+			count = count / replica
```
Just highlighting this hunk to @jarrpa 's attention - this was discussed very controversally before, e.g. #835  ... :smile:

jarrpa replied:
Yes, don't do this. We want to freeze this behavior as hard as possible.

raghavendra-talur replied:
Done

obnoxxx commented on pkg/controller/storagecluster/reconcile.go:
```diff
@@ -697,3 +701,13 @@ func newCleanupJob(sc *ocsv1.StorageCluster) *batchv1.Job {
+func validateArbiterSpec(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+
+	if sc.Spec.Arbiter.Enable {
+		if sc.Spec.NodeTopologies.ArbiterLocation == "" {
```
done

iamniting replied:
Done

obnoxxx commented on pkg/controller/storagecluster/reconcile.go:
```diff
@@ -697,3 +701,13 @@ func newCleanupJob(sc *ocsv1.StorageCluster) *batchv1.Job {
+
+	if sc.Spec.Arbiter.Enable {
+		if sc.Spec.NodeTopologies.ArbiterLocation == "" {
+			return fmt.Errorf("arbiter is set to enable but no arbiterLocation has been provided")
```
done

iamniting replied:
Done

jarrpa commented on pkg/controller/storagecluster/cephcluster.go:
```diff
@@ -343,8 +340,12 @@ func newExternalCephCluster(sc *ocsv1.StorageCluster, cephImage string, monitori
+}
+
 func getMonCount(nodeCount int) int {
-	count := defaults.MonCountMin
```
This removes the only reference to this variable in the code. Either keep this logic, or remove the MonCountMin variable.

raghavendra-talur replied:
Done

jarrpa commented on pkg/controller/storagecluster/topology.go:
```diff
@@ -219,7 +219,7 @@ func (r *ReconcileStorageCluster) ensureNodeRacks(
 // in the storage cluster
 func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
-	minNodes := getMinNodes(sc)
+	minNodes := getMinimumNodes(sc)
```
Unnecessary function rename?

raghavendra-talur replied:
This was a function that was added in a different commit of the same PR. I have updated the commits to not have this function rename.

jarrpa commented on hack/functest.sh:
```diff
@@ -13,3 +13,16 @@ if [ $? -ne 0 ]; then
 	echo "ERROR: Functest failed."
 	exit 1
 fi
+
+"$OUTDIR_BIN/${GINKGO_TEST_SUITE}_tests" -ginkgo.v \
+        --ocs-registry-image="${CATALOG_FULL_IMAGE_NAME}" \
+	--ocs-subscription-channel="${OCS_SUBSCRIPTION_CHANNEL}" \
+	--upgrade-from-ocs-registry-image="${UPGRADE_FROM_OCS_REGISTRY_IMAGE}" \
+	--upgrade-from-ocs-subscription-channel="${UPGRADE_FROM_OCS_SUBSCRIPTION_CHANNEL}" \
+	--ocs-cluster-uninstall="${OCS_CLUSTER_UNINSTALL}" \
+	--arbiter "$@"
+
+if [ $? -ne 0 ]; then
+	echo "ERROR: Arbiter mode functest failed."
+	exit 1
+fi
```
I don't want to do this, we are in effect doubling the length of the functional tests. We need to think about how we want to run this... for now I would just say to not add this.

raghavendra-talur replied:
Done

jarrpa reviewed (changes requested):
In addition to the comments below, I see a number of places where there are unnecessary newlines in functions. It's a matter of taste, to be sure, but I don't like blank lines at the beginning and end of functions. If possible I'd like to remove the blank lines in the commits they are introduced, but if it's too much of a hassle I would accept a single formatting commit at the end of the PR.

umangachapagain reviewed (changes requested):
I'd like to see all the Rook/Ceph image and dependency changes at one place and compact. Probably a separate PR for it just like we did with NooBaa changes for KMS.

With these changes we're looking at some unsupported, custom image references (as the spec `AllowUnsupported: true` rightly says). It'd be safer to test this separately.

There's a CEPH_IMAGE env var in the CSV which didn't change (it should).
There's some unwanted changes in a commit that upgrades Rook.

Does it sound reasonable? @obnoxxx @jarrpa

raghavendra-talur commented:
> I'd like to see all the Rook/Ceph image and dependency changes at one place and compact. Probably a separate PR for it just like we did with NooBaa changes for KMS.

Makes sense. Created #964 
> 
> With these changes we're looking at some unsupported, custom image references (as the spec `AllowUnsupported: true` rightly says). It'd be safer to test this separately.

This was introduced for testing. I have removed it now.

> 
> There's a CEPH_IMAGE env var in the CSV which didn't change (it should).
> There's some unwanted changes in a commit that upgrades Rook.
This is also updated in #964 now.

> 
> Does it sound reasonable? @obnoxxx @jarrpa

I will rebase this PR once #964 is merged.

obnoxxx commented:
> > I'd like to see all the Rook/Ceph image and dependency changes at one place and compact. Probably a separate PR for it just like we did with NooBaa changes for KMS.
> 
> Makes sense. Created #964
> 
> > With these changes we're looking at some unsupported, custom image references (as the spec `AllowUnsupported: true` rightly says). It'd be safer to test this separately.
> 
> This was introduced for testing. I have removed it now.
> 
> > There's a CEPH_IMAGE env var in the CSV which didn't change (it should).
> > There's some unwanted changes in a commit that upgrades Rook.
> > This is also updated in #964 now.
> 
> > Does it sound reasonable? @obnoxxx @jarrpa

Sounds good, @umangachapagain !

> I will rebase this PR once #964 is merged.

You can already rebase now, if #964 merges as is, this won't need to be rebased again (merge will take care of it). @raghavendra-talur

jarrpa commented:
  | /test ocs-operator-ci
-- | --

jarrpa commented:
/test ocs-operator-ci

jarrpa commented on pkg/controller/storagecluster/cephblockpools.go:
```diff
@@ -14,15 +14,24 @@ import (
 		Size:            3,
 		TargetSizeRatio: .49,
 	}
+
```
Remove

jarrpa commented on pkg/controller/storagecluster/cephblockpools.go:
```diff
@@ -14,15 +14,24 @@ import (
 // newCephBlockPoolInstances returns the cephBlockPool instances that should be created
 // on first run.
 func (r *ReconcileStorageCluster) newCephBlockPoolInstances(initData *ocsv1.StorageCluster) ([]*cephv1.CephBlockPool, error) {
+
```
Remove

jarrpa commented on pkg/controller/storagecluster/storagecluster_controller_test.go:
```diff
@@ -3,9 +3,10 @@ package storagecluster
-	consolev1 "github.com/openshift/api/console/v1"
 	"testing"
 
+	consolev1 "github.com/openshift/api/console/v1"
+
```
Is this auto-formatting? Either way it should go with the other OCP imports below.

jarrpa reviewed (changes requested):
I think overall the code looks good, and the commits are clean. A few cosmetic changes are requested.

jarrpa reviewed (approved):
/lgtm
